File: Projects/SteinsGate-Common.Core/Sources/SteinsGate/Common/SgaSound.cpp
/*
 * 작성자: 윤정도
 * 생성일: 1/8/2023 6:25:59 AM
 * =====================
 *
 */

#include "Core.h"
#include <SteinsGate/Common/SgaSound.h>
#include <SteinsGate/Common/SgaPackage.h>

USING_NS_JC;

//////////////////////////////////////////////////////////////////////////////////////////
SgaSound::~SgaSound()
{
}

//////////////////////////////////////////////////////////////////////////////////////////
bool SgaSound::Load(bool _indexOnly)
{
	if (_indexOnly)
		return true;

	if (!pParent_.Exist())
		return false;

	pData_ = MakeShared<Byte[]>(indexLength_);
	pParent_->StreamRef().Seek(indexOffset_);
	pParent_->StreamRef().Read(pData_.GetPtr(), 0, pData_.Length());
	return true;
}

//////////////////////////////////////////////////////////////////////////////////////////
bool SgaSound::Unload()
{
	if (!pData_.Exist())
		return false;

	pData_ = nullptr;
	return true;
}

//////////////////////////////////////////////////////////////////////////////////////////
SgaSoundPtr SgaSound::Create(const Header& _header)
{
	return MakeShared<SgaSound>(_header);
}

File: Projects/SteinsGate-Tools.Common/Project/Sga/SgaSound.cs
/*
 * 작성자: 윤정도
 * 생성일: 2/27/2023 2:52:49 AM
 *
 */

using System.IO;

namespace SGToolsCommon.Sga
{
    public class SgaSound : SgaElement
    {
        private byte[] data_;

        public byte[] Data => data_;

        //////////////////////////////////////////////////////////////////////////////////
        public SgaSound(SgaPackage _parent, SgaElementHeader _header, int _version, int _indexOffset, int _indexLength)
            : base(SgaElementType.Sound, _parent, _header, _version, _indexOffset, _indexLength)
        {
            data_ = null;
            indexLoaded_ = true;
            dataLoaded_ = false;
        }

        //////////////////////////////////////////////////////////////////////////////////
        public override void Load(bool _indexOnly)
        {
            if (_indexOnly)
                return;

            data_ = new byte[IndexLength];
            dataLoaded_ = true;
            parent_.ReadStream.Seek(IndexOffset, SeekOrigin.Begin);
            parent_.ReadStream.Read(data_, 0, IndexLength);
        }

        //////////////////////////////////////////////////////////////////////////////////
        public override void Unload()
        {
            data_ = null;
            dataLoaded_ = false;
        }
    }
}
